src/ast.erl

```diff
@@ -467,40 +475,195 @@ is_predef_alias_name(N) ->
 % smart constructors for intersection, union and negation
 -spec mk_intersection([ast:ty()]) -> ast:ty().
 mk_intersection(Tys) ->
-    Filtered =
-        lists:filter(
-          fun(T) ->
-                  case T of
-                      {predef, any} -> false;
-                      {negation, {predef, none}} -> false;
-                      _ -> true
-                  end
-          end,
-          Tys),
-    case Filtered of
-        [] -> {predef, any};
-        [T] -> T;
-        _ -> {intersection, Filtered}
+    HasEmpty =
+        lists:any(
+            fun(T) ->
+                case T of
+                    {predef, none} -> true;
+                    {negation, {predef, any}} -> true;
+                    _ -> false
+                end
+            end,
+            Tys),
+    case HasEmpty of
+        true -> {predef, none};
+        _ ->
+            Filtered =
+                lists:filter(
+                    fun(T) ->
+                        case T of
+                            {predef, any} -> false;
+                            {negation, {predef, none}} -> false;
+                            _ -> true
+                        end
+                    end,
+                    Tys),
+            case Filtered of
+                [] -> {predef, any};
+                [T] -> T;
+                _ -> {intersection, Filtered}
+            end
     end.
 
+
 -spec mk_union([ast:ty()]) -> ast:ty().
 mk_union(Tys) ->
-    Filtered =
-        lists:filter(
-          fun(T) ->
-                  case T of
-                      {predef, none} -> false;
-                      _ -> true
-                  end
-          end,
-          Tys),
-    case Filtered of
-        [] -> {predef, none};
-        [T] -> T;
-        _ -> {union, Filtered}
+    HasAny =
+        lists:any(
+            fun(T) ->
+                case T of
+                    {predef, any} -> true;
+                    {negation, {predef, none}} -> true;
+                    _ -> false
+                end
+            end,
+            Tys),
+    case HasAny of
+        true -> {predef, any};
+        _ ->
+            Filtered =
+                lists:filter(
+                    fun(T) ->
+                        case T of
+                            {predef, none} -> false;
+                            [] -> error(todo);
+                            _ -> true
+                        end
+                    end,
+                    Tys),
+            case Filtered of
+                [] -> {predef, none};
+                [T] -> T;
+                _ -> {union, Filtered}
+            end
     end.
 
 -spec mk_negation(ast:ty()) -> ast:ty().
 mk_negation({predef, any}) -> {predef, none};
 mk_negation({predef, none}) -> {predef, any};
 mk_negation(T) -> {negation, T}.
+
+erlang_ty_to_ast(X) ->
```

Comment:
Imo, this function should be moved somewhere else. For me, the ast module is a specification and should not contain much code. I already thought about moving the smart constructors elsewhere.

Reply:
Will be split up into `ast.erl`  and  `ast_lib.erl`